### ISSUE TITLE
about project_2d_sym

### ISSUE THREAD

**Issue by joyjoyzyz** [zju3dv/clean-pvnet#165]
hello, thanks for your sharing code.
I have read #130 , and modify the code lib/evaluators/linemod/pvnet.py, looks as follow:

    def projection_2d(self, pose_pred, pose_targets, K, icp=False, syn=False, threshold=5):
        model_2d_pred = pvnet_pose_utils.project(self.model, K, pose_pred)
        model_2d_targets = pvnet_pose_utils.project(self.model, K, pose_targets)
        #proj_mean_diff = np.mean(np.linalg.norm(model_2d_pred - model_2d_targets, axis=-1))
        
        if syn:
            idxs = nn_utils.find_nearest_point_idx(model_2d_pred, model_2d_targets)
            proj_mean_diff = np.mean(np.linalg.norm(model_2d_pred[idxs] - model_2d_targets, 2, 1))
        else:
            proj_mean_diff = np.mean(np.linalg.norm(model_2d_pred - model_2d_targets, axis=-1))
        
        if icp:
            self.icp_proj2d.append(proj_mean_diff < threshold)
        else:
            self.proj2d.append(proj_mean_diff < threshold)

I don't know if it's a right modification, because i got strange results 2d projections results  after evaluate on "eggbox" and "glue":
Occlusion_linemod eggbox：
2d projections metric: 0.8238297872340425
ADD metric: 0.44340425531914895
5 cm 5 degree metric: 0.007659574468085106
mask ap70: 0.5140425531914894

Occlusion_linemod glue：
2d projections metric: 0.8560354374307863
ADD metric: 0.4097452934662237
5 cm 5 degree metric: 0.20376522702104097
mask ap70: 0.6843853820598007
Could you add this part of code? because I want to use project_2d_sym() to evaluate on symmetric objects, like ADD metirc.
thanks in advance.

I have read it https://github.com/zju3dv/pvnet/blob/master/lib/utils/evaluation_utils.py#L83, and modified as it, but get strange results, so I ask your advice, if my modification is right.

**Comment by joyjoyzyz** [comment]
I don't know why the issue was closed, I just need your advice, please.

**Comment by pengsida** [comment]
Why do you say the results are strange?

**Comment by joyjoyzyz** [comment]
sorry , I said the 2d projections results  is strange, because in the paper is 
Occlusion_linemod eggbox：
2d projections metric: 8.43 
the modified project_2d_sym() got 82.38

**Comment by pengsida** [comment]
2D projection metric is not calculated using `project_2d_sym()`.

**Comment by joyjoyzyz** [comment]
I modify the code lib/evaluators/linemod/pvnet.py, and add a "if", just like add_metric()
def projection_2d(self, pose_pred, pose_targets, K, icp=False, syn=False, threshold=5):
    model_2d_pred = pvnet_pose_utils.project(self.model, K, pose_pred)
    model_2d_targets = pvnet_pose_utils.project(self.model, K, pose_targets)
    #proj_mean_diff = np.mean(np.linalg.norm(model_2d_pred - model_2d_targets, axis=-1))
    
    if syn:
        idxs = nn_utils.find_nearest_point_idx(model_2d_pred, model_2d_targets)
        proj_mean_diff = np.mean(np.linalg.norm(model_2d_pred[idxs] - model_2d_targets, 2, 1))
    else:
        proj_mean_diff = np.mean(np.linalg.norm(model_2d_pred - model_2d_targets, axis=-1))
    
    if icp:
        self.icp_proj2d.append(proj_mean_diff < threshold)
    else:
        self.proj2d.append(proj_mean_diff < threshold)

if cfg.cls_type in ['eggbox', 'glue']:
        self.projection_2d(pose_pred, pose_gt, syn=True)
        self.add_metric(pose_pred, pose_gt, syn=True)

**Comment by pengsida** [comment]
For 2D projection metric, eggbox and glue are not considered as symmetric objects.

**Comment by joyjoyzyz** [comment]
in #130 you said "The eggbox is taken as symmetric for all metrics." I think I am confused

**Comment by pengsida** [comment]
Sorry. I have revised this reply: https://github.com/zju3dv/clean-pvnet/issues/130#issuecomment-689342977

**Comment by joyjoyzyz** [comment]
sorry, I have another confusion, In the code "pvnet-master", there is a https://github.com/zju3dv/pvnet/blob/master/lib/utils/evaluation_utils.py#L83 project_2d_sym(), and 
if class_type in ['eggbox','glue']:
        self.projection_2d_sym(pose_pred, pose_targets, model, self.projector.intrinsic_matrix[intri_type])
        self.add_metric_sym(pose_pred, pose_targets, model, diameter)
else:
        self.projection_2d(pose_pred, pose_targets, model, self.projector.intrinsic_matrix[intri_type])
        self.add_metric(pose_pred, pose_targets, model, diameter)

**Comment by pengsida** [comment]
As shown in paper, eggbox and glue are not considered as symmetric objects for the 2d projection metric.

**Comment by joyjoyzyz** [comment]
ok, thank you. I open this issue because I saw someone have lower 2d_projection scores than paper, https://github.com/zju3dv/pvnet/issues/91
and I also got lower 2d_projection scores on occlusion linemod dataset, the scores under other metirc is similar to the paper.

**Comment by pengsida** [comment]
Do you use the released pre-trained models?

**Comment by joyjoyzyz** [comment]
yes, I use the pre-trained models to test, and got 59.2 average 2d_projection scores, where 61.06 in paper.
like eggbox 5.36(paper:8.43), can:83.02(paper:86.09), holepuncher:65.77(paper:69.84) ...

**Comment by pengsida** [comment]
The performance gap is very small.
The results of ransac voting depend on the random seed.
Previously we do not fix the random seed.
In this code, we fix the random seed here: https://github.com/zju3dv/clean-pvnet/blob/master/run.py#L67

**Comment by joyjoyzyz** [comment]
thanks
you mean the result 59.2 is normal, is it right?

**Comment by pengsida** [comment]
Yes.
